[Merged by Bors] - chore: replace open scoped Classical with `open scoped Classical in` or `classical`

kvanvels opened the pull request:
reduces some technical debt by replacing "open scoped classical" with (open scoped classical in) for theorem declarations and/or adding a "classical' tactic to a tactic proof.  I followed the example in PR  #20325.  

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sgouezel commented:
Thanks for doing this! As you can see, I just have minor comments.

kvanvels commented:
[sgouezel](https://github.com/sgouezel) wrote:

> Thanks for doing this! As you can see, I just have minor comments.

I don't see any comments.  Am I missing something obvious?

sgouezel commented:
Weird, github had sent my conclusion message, but not the individual comments on the code. I have tried again, hopefully they have landed now.

kvanvels commented:
I got the comments.  I will push the corrections tonight or tomorrow.

I was wondering about limiting the scope of the "classicals".  I put them as far down the proofs as I was able, (I think).  So that way future people can deduce where classical reasoning is required.  Your thoughts?

sgouezel commented:
Personally, I would put it at the start of the proof for simplicity (it's not that it enables classical reasoning, which is always enabled in fact, it's that it provides decidability assumptions for some lemmas that need it, so it's mostly a technical implementation detail IMHO). But putting it in the middle of the proof as you did is also fine.

kvanvels commented:
Hmmm, ok.  So I was mixing up classical reasoning (the axiom of choice and its consequences) and decidability/computability?

kvanvels commented on Mathlib/Probability/Process/HittingTime.lean:
```diff
@@ -47,10 +47,15 @@ variable {Ω β ι : Type*} {m : MeasurableSpace Ω}
-noncomputable def hitting [Preorder ι] [InfSet ι] (u : ι → Ω → β) (s : Set β) (n m : ι) : Ω → ι :=
-  fun x => if ∃ j ∈ Set.Icc n m, u j x ∈ s then sInf (Set.Icc n m ∩ {i : ι | u i x ∈ s}) else m
+noncomputable def hitting [Preorder ι] [InfSet ι] (u : ι → Ω → β)
+(s : Set β) (n m : ι) : Ω → ι := by
```
I think I fixed this in a merge commit. I will double check this, but you probably should too.  I had to indent some subsequent lines.

kvanvels commented on Mathlib/RingTheory/Adjoin/Tower.lean:
```diff
@@ -58,17 +58,20 @@ end Algebra
-    rw [Finset.coe_union, Finset.coe_image, Algebra.adjoin_algebraMap_image_union_eq_adjoin_adjoin,
-      hs, Algebra.adjoin_top, ht, Subalgebra.restrictScalars_top, Subalgebra.restrictScalars_top]⟩
-
+  by
```
Fixed in a separate commit.

kvanvels commented:
I think committed at faux pas.  I did some more work in removing some "Classical"s this morning (in new files) and I added them to this pull request. I modified some pretty basic (close to the root) files and my computer was taking forever to compile them.  I pushed them here (without thinking it through) to have the servers here compile them.

Should I make a new branch to continue fixing these "classical"s?

sgouezel commented:
It's not really a faux pas, but the PR is reaching a critical side, so I think it is a good point to stop there and wait for a review. If you want to go on working on this, you can do it on another branch (either based on this one, or from master).

sgouezel commented:
You can also see that there is an "Awaiting-author" tag on the PR. This means that the PR will not show up on the reviewers dashboard currently. So, once you are happy with the state of your PR, it is important that you remove this tag, to indicate that your PR is ready for review.

kvanvels commented:
There was one final request from your initial review that I just fixed.  Am I correct in assuming that that is what removed the "awaiting-author" tag on the PR?

j-loreaux commented:
No, *you* removed the `awaiting-author` tag (probably by clicking on it). This is fine if and only if you're ready for the PR to be reviewed again, which it seems is the case because all the conversations are marked as resolved.

j-loreaux commented on Archive/Wiedijk100Theorems/Partition.lean:
```diff
@@ -93,9 +91,11 @@ universe u
 /-- A convenience constructor for the power series whose coefficients indicate a subset. -/
-def indicatorSeries (α : Type*) [Semiring α] (s : Set ℕ) : PowerSeries α :=
-  PowerSeries.mk fun n => if n ∈ s then 1 else 0
+def indicatorSeries (α : Type*) [Semiring α] (s : Set ℕ) : PowerSeries α := by
+  classical
+  exact PowerSeries.mk fun n => if n ∈ s then 1 else 0
```
I would still write this as:
```suggestion
open scoped Classical in
def indicatorSeries (α : Type*) [Semiring α] (s : Set ℕ) : PowerSeries α :=
  PowerSeries.mk fun n => if n ∈ s then 1 else 0
```
although this suggestion won't work as is because `open scoped Classical in` must go before the docstring.

The reason I'm suggesting this is because it's nice to avoid doing `:= by` when writing `def`s (since it's just `classical; exact ...` it should be fine this way, but it's best to encourage good habits).

kvanvels replied:
That makes sense.  I suspect there are quite a few of those.

j-loreaux commented on Mathlib/Computability/TuringMachine.lean:
```diff
@@ -1205,8 +1205,9 @@ theorem stmts₁_supportsStmt_mono {S : Finset Λ} {q₁ q₂ : Stmt₁} (h : q
 halt state. This is used in the TM1 to TM0 reduction. -/
-noncomputable def stmts (M : Λ → Stmt₁) (S : Finset Λ) : Finset (Option Stmt₁) :=
-  Finset.insertNone (S.biUnion fun q ↦ stmts₁ (M q))
+noncomputable def stmts (M : Λ → Stmt₁) (S : Finset Λ) : Finset (Option Stmt₁) := by
+  classical
+  exact Finset.insertNone (S.biUnion fun q ↦ stmts₁ (M q))
```
idem

j-loreaux commented on Mathlib/Computability/TuringMachine.lean:
```diff
@@ -1710,9 +1710,11 @@ noncomputable def writes : Stmt₁ → Finset Λ'₁
 are the normal states embedded from `S`, plus all write states accessible from these states. -/
-noncomputable def trSupp (S : Finset Λ) : Finset Λ'₁ :=
-  S.biUnion fun l ↦ insert (Λ'.normal l) (writes (M l))
+noncomputable def trSupp (S : Finset Λ) : Finset Λ'₁ := by
+  classical
+  exact S.biUnion fun l ↦ insert (Λ'.normal l) (writes (M l))
```
idem

j-loreaux commented on Mathlib/Computability/TuringMachine.lean:
```diff
@@ -2005,8 +2008,9 @@ theorem stmts₁_supportsStmt_mono {S : Finset Λ} {q₁ q₂ : Stmt₂} (h : q
 /-- The set of statements accessible from initial set `S` of labels. -/
-noncomputable def stmts (M : Λ → Stmt₂) (S : Finset Λ) : Finset (Option Stmt₂) :=
-  Finset.insertNone (S.biUnion fun q ↦ stmts₁ (M q))
+noncomputable def stmts (M : Λ → Stmt₂) (S : Finset Λ) : Finset (Option Stmt₂) := by
+  classical
+  exact Finset.insertNone (S.biUnion fun q ↦ stmts₁ (M q))
```
idem

j-loreaux commented on Mathlib/Probability/ProbabilityMassFunction/Constructions.lean:
```diff
@@ -82,6 +82,7 @@ variable (s : Set β)
 @[simp]
 theorem toOuterMeasure_map_apply : (p.map f).toOuterMeasure s = p.toOuterMeasure (f ⁻¹' s) := by
   simp [map, Set.indicator, toOuterMeasure_apply p (f ⁻¹' s)]
+  rfl
```
Why is this needed now?

kvanvels replied:
To be honest I don't understand why.  IIRC, the goal was Something = Something.

kvanvels replied:
I double checked.  The goal before the rfl line is
⊢ (∑' (a : α), if f a ∈ s then p a else 0) = ∑' (x : α), if x ∈ f ⁻¹' s then p x else 0

I don't know why this wasn't required before an now it is.

j-loreaux commented on Mathlib/Probability/Process/HittingTime.lean:
```diff
@@ -47,10 +47,15 @@ variable {Ω β ι : Type*} {m : MeasurableSpace Ω}
 The hitting time is a stopping time if the process is adapted and discrete. -/
-noncomputable def hitting [Preorder ι] [InfSet ι] (u : ι → Ω → β) (s : Set β) (n m : ι) : Ω → ι :=
-  fun x => if ∃ j ∈ Set.Icc n m, u j x ∈ s then sInf (Set.Icc n m ∩ {i : ι | u i x ∈ s}) else m
+noncomputable def hitting [Preorder ι] [InfSet ι] (u : ι → Ω → β)
+  (s : Set β) (n m : ι) : Ω → ι := by
+    classical
+    exact
+      fun x => if ∃ j ∈ Set.Icc n m, u j x ∈ s
+               then sInf (Set.Icc n m ∩ {i : ι | u i x ∈ s}) else m
```
same comment about `def := by classical ...`

j-loreaux reviewed (commented):
some pretty minor comments.

jcommelin commented on Mathlib/CategoryTheory/Preadditive/Mat.lean:
```diff
@@ -94,12 +93,17 @@ section
 
 attribute [local simp] Hom.id Hom.comp
 
+
```
```suggestion
```

jcommelin commented on Mathlib/CategoryTheory/Preadditive/Mat.lean:
```diff
@@ -244,11 +251,13 @@ def mapMat_ (F : C ⥤ D) [Functor.Additive F] : Mat_ C ⥤ Mat_ D where
   obj M := ⟨M.ι, fun i => F.obj (M.X i)⟩
   map f i j := F.map (f i j)
 
+
```
```suggestion
```

jcommelin commented on Mathlib/Probability/Process/Stopping.lean:
```diff
@@ -96,7 +96,9 @@ protected theorem measurableSet_eq_of_countable_range (hτ : IsStoppingTime f τ
   refine MeasurableSet.biUnion h_countable fun j _ => ?_
+  classical
   rw [Set.iUnion_eq_if]
+
```
```suggestion
```

jcommelin commented on Mathlib/RingTheory/Adjoin/FG.lean:
```diff
@@ -129,11 +129,12 @@ theorem FG.prod {S : Subalgebra R A} {T : Subalgebra R B} (hS : S.FG) (hT : T.FG
-open scoped Classical
 
-theorem FG.map {S : Subalgebra R A} (f : A →ₐ[R] B) (hs : S.FG) : (S.map f).FG :=
+
```
```suggestion
```

kvanvels replied:
There was three blank lines here.  I deleted two of them.

jcommelin reviewed (commented):
There are a few empty lines in the diff. Please remove them.

Otherwise, this looks good to me.

jcommelin commented:
Please fix the merge conflict by merging `master`

kvanvels commented:
I am unsure how to resolve the branch conflicts.  It seems  
Mathlib/RingTheory/MaximalSpectrum.lean
and Mathlib/RingTheory/PrimeSpectrum.lean 
have both been deleted.  Is this something that I would do or someone higher up?

kvanvels commented:
J commelin, I didn't see your message when I posted mine.  Sorry.  I think I can figure it out.  I will write back here if I can't.

jcommelin commented:
Both files moved to `Mathlib/RingTheory/Spectrum/{Prime,Maximal}/{Defs/Basic}.lean`.

kvanvels commented:
@j-loreaux  and @jcommelin, I think I screwed this up somehow.  My merge undid someone else's work in Mathlib/RingTheory/AlgebraTower.lean.  Can either of you explain how I can discard all the changes I made to that file?  I can take care of the classicals at a later PR. I am now getting errors in /Mathlib/LinearAlgebra/Matrix/ToLin.lean (a file I don't think I touched).

kim-em commented:
> @j-loreaux and @jcommelin, I think I screwed this up somehow. My merge undid someone else's work in Mathlib/RingTheory/AlgebraTower.lean. Can either of you explain how I can discard all the changes I made to that file? I can take care of the classicals at a later PR. I am now getting errors in /Mathlib/LinearAlgebra/Matrix/ToLin.lean (a file I don't think I touched).

A simple approach is:
```
git fetch
git checkout origin/master Mathlib/RingTheory/AlgebraTower.lean
git add Mathlib/RingTheory/AlgebraTower.lean
git commit -am "revert changes to Mathlib/RingTheory/AlgebraTower.lean"
```

kvanvels commented:
@kim-em, thank you.  I appreciate it.

kvanvels commented:
I am making this comment to bump this PR and to see if  I need to do anything else.

jcommelin commented:
@kvanvels It looks like there is another merge conflict. Sorry about that.
Please try to fix it, and afterwards merge the PR by instructing our bot: `bors r+`.

bors d+

kvanvels commented:
bors r+
